management/lifecycle.py

```diff
@@ -26,6 +27,122 @@
+    """Try to extract a SMILES from multiple fields in ``entry``. Based on code
+    from qcsubmit, see ``BasicResultCollection.from_datasets``, for example.
+    """
+    # this should cover either single-points or optimizations
+    molecule = (
+        getattr(entry, "molecule", None)
+        or getattr(entry, "initial_molecule", None)
+    )
+    # this is how qcsubmit gets a molecule for torsion drives
+    if molecule is None:
+        return entry.attributes.get(
+            "canonical_isomeric_explicit_hydrogen_mapped_smiles"
+        )
+    # and this covers all the possibilities for finding a smiles in
+    # single-points or optimizations
+    return (
+        molecule.identifiers.canonical_isomeric_explicit_hydrogen_mapped_smiles
+        or molecule.extras.get(
+            "canonical_isomeric_explicit_hydrogen_mapped_smiles"
+        )
+        or entry.attributes.get(
+            "canonical_isomeric_explicit_hydrogen_mapped_smiles"
+        )
+    )
+
+
+def partition_records(ds, bins, include_complete=False) -> dict[int, list[int]]:
+    """Split up the records in ``ds`` based on the molecular weights (in Da) in
+    ``bins``.
+
+    ``include_complete`` is intended mostly (only?) for testing; usually you
+    wouldn't want to retag complete records (and it might be an error in
+    qcportal since complete records don't have tags), but for tests it's nice
+    to be able to call this on a finished dataset
+    """
+    from openff.toolkit import Molecule
+    from openff.units import unit
+
+    # mapping of bin index to a sequence of record_ids
+    ret: dict[int, list[int]] = defaultdict(list)
+
+    ds.fetch_entries()
+
+    for entry_name, _, rec in ds.iterate_records():
+        if rec.status == "complete" and not include_complete:
+            continue
+        entry = ds.get_entry(entry_name)
+        if (cmiles := try_get_smiles(entry)) is None:
+            print(f"failed to get cmiles from {entry_name}")
+            continue
+        mol = Molecule.from_mapped_smiles(cmiles, allow_undefined_stereo=True)
+        mass = sum(atom.mass for atom in mol.atoms)
+
+        # iterate through the sequence of bins, putting the record id in the
+        # bin if its mass is less than the bin threshold. use i+1 after the
+        # loop for the largest records (above any threshold)
+        found = False
+        for i, threshold in enumerate(bins):
+            if mass < threshold * unit.dalton:
+                ret[i].append(rec.id)
+                found = True
+                break
+
+        if not found:
+            ret[i+1].append(rec.id)
+
+    return ret
+
+
+def set_mw_compute_tags(client, ds, compute_tag, include_complete=False):
+    bins, base_tag = parse_tags(compute_tag)
+    if len(bins) == 0:
+        print(f"Failed to parse molecular weight compute tags from {compute_tag}")
+        # TODO should this fall back on the normal setter then? we'd need to
+        # pass more arguments to make that possible
```

Comment:
Possibly I should add this to the function docstring, but this function is only supposed to be called when `SPLIT_TAG` matches (see `update_compute_tags`), so this `len(bins)` check *should* be overly cautious anyway. However, I thought it was still good to leave some kind of check in case of future changes.